init blog post re exceptions (#32)
sampsyo/cs6120: content/blog/2019-09-25-exceptions-in-bril.md

+++
title = "Exceptions in Bril"
extra.author = "Rolph Recto"
extra.bio = """
[Rolph Recto](rolph-recto.github.io) is a third-year graduate student studying
the intersection of programming languages, security, and distributed systems.
"""
+++

## Overview

We add exceptions as a control structure in Bril to support non-local
control flow. 
This allows frontends to compile exceptions directly.
The ultimate goal is to extend Bril further with more powerful control
structures such as first-class continuations or algebraic effects.

## Function Calls

To make the implementation of exceptions nontrivial and to make exceptions
actually be non-local control structures, we first extend Bril so that it
supports function calls.
We do this by adding a `call` instruction that takes as arguments a
function name and a list of function arguments.
We also extended function declarations to allow for formal parameters.
The Bril text format now allows for declarations of the form

```
funcName arg1:type1 ... argN:typeN { instrs }
```

> Our implemention explicitly represents the program stack as an object. 

Reviewer: The interpreter implementation, specifically?

> The interpreter has four configuration variables that consist of its current

Reviewer: ```suggestion
The interpreter has four configuration variables that make up its current
```

That is, the AR consists of four parts, rather than the other way around?

Reviewer: Also, I might suggest not saying "configuration variables"—I know what you're going for here (a machine "configuration") but it sounds like the interpreter has a config file.

Author: Thanks, that does sound a little stilted!

activation record:

* Bril function object --- the current function whose list of instructions
  is being executed

* Variable environment --- map from local variables to values

> * Handler environment --- map from exception handlers to (see below)

Reviewer: A map from exception names to handlers?

Author: Yup! Thanks for the catch

* PC index --- the index into the instruction list of the current instruction
  being executed

The interpreter proceeds by executing instructions, which either change

> environment mappings and/or sets the PC index of the next instruction

Reviewer: ```suggestion
environment mappings and/or set the PC index of the next instruction
```

Also, set the PC _to_ the index of the next instruction, perhaps?

to execute.
When `call` is executed, the interpreter does the following:

* The current activation record is pushed into the program stack.

* An empty variable environment is constructed, and the formal parameters
  of the function are mapped to the arguments of the `call` instruction.
  An empty handler environment is constructed.

* The PC index is changed to 0, and the variable and handler environments 

>   are updated.

Reviewer: How are the variable and handler environments updated? (I thought we just initialized them in the previous step?)

Author: I meant that the environments were emptied since the function call creates a new activation record.

  a exception (in the metalanguage) since a Bril exception was not handled
  properly.

* Once it finds an appropriate handler, the interpreter sets the
  handler's activation record as its current activation record and then
  it jumps to the PC index of the handler label.

## Apologia

Our implementation explicitly represents the program stack.
Our implementation could also have implicitly represented the program stack
using the interpreter's own stack by making recursive calls to 
`evalFunc`, but we obviated this design for the more explicit one to have
finer control over the program stack of Bril and to support
future development of other non-local control structures that
manipulate the stack.
(For example, first-class continuations would involve storing program stacks
as values in the environment map.)
Also, the implicit representation of program stacks would essentially fix
the non-local control structures of Bril to the control structures
the metalanguage has.
With the implicit representation, to throw exceptions in Bril programs
one would need to throw an exception in the interpreter as well;
supporting first-class continuations or algebraic effects would be impossible

> in this way.

Reviewer: This paragraph makes a very good point!

Our implementation of exception handlers does not support passing
exception objects.

> To support this in the future, we make handlers be standalone functions instead

Reviewer: ```suggestion
To support this in the future, we will make handlers be standalone functions instead
```

Maybe? In any case, there's a missing verb.

of just labels in an existing function.
This would make reasoning about the control flow of the program even less local
since it would make handlers not syntactically part of the contexts in which
they might be thrown, so we obviated this design choice for the current
implementation.

Another limitation of the current implementation of exceptions is the fact that
handlers are not tied to a lexical scope. 
That means a handler would override any handler for the same exception name
installed at the same function.
We choose this simpler albeit somewhat unintuitive design to obviate the
introduction of nested lexical scopes in Bril, which we believe is

> antithetical to its intended use as a simple intermediate language.

Reviewer: Indeed; it seems fine here—the source language can enforce this constraint.

## Evaluation

Our main goal for evaluation is to check whether our implementation is correct.
We considered performance considerations as out of bounds for evaluation since
the Bril interpreter is sufficiently different enough from how, say, ISA
instructions are implemented in hardware that we cannot infer anything useful
regarding the performance of an analogous implemenation of our exceptions

> mechanism in the latter from results in the former.

Reviewer: Yep, I buy it.
